src/main/scala/strawman/collection/mutable/ArrayDeque.scala

```diff
@@ -0,0 +1,349 @@
+          elems.copyToArray(array2.asInstanceOf[Array[A]], idx)
+          copySliceToArray(srcStart = idx, dest = array2, destStart = idx + srcLength, maxItems = size)
+          set(array = array2, start = 0, end = finalLength)
+        } else {
+          val suffix = drop(idx)
+          end = start_+(idx)
+          elems.foreach(appendAssumingCapacity)
+          suffix.foreach(appendAssumingCapacity)
+        }
+      case _ => //expensive to compute size
+        val suffix = drop(idx)
+        end = start_+(idx)
+        this ++= elems ++= suffix
+    }
+  }
+
+  override def remove(idx: Int, count: Int): Unit = {
+    ArrayDeque.checkIndex(idx, this)
+    if (count <= 0) return
+    val removals = (size - idx) min count
+    // If we are removing more than half the elements, its cheaper to start over
+    // Else, choose the shorter: either move the prefix (0 until (idx + removals) right OR the suffix (idx to size) left
+    if (2*removals >= size) {
+      val array2 = ArrayDeque.alloc(size - removals)
+      copySliceToArray(srcStart = 0, dest = array2, destStart = 0, maxItems = idx)
+      copySliceToArray(srcStart = idx + removals - 1, dest = array2, destStart = idx, maxItems = size)
+      set(array = array2, start = 0, end = size - removals)
+    } else if (size - idx <= idx + removals) {
+      var i = idx
+      while(i + removals < size) {
+        set(i, get(i + removals))
+        set(i + removals, null)
+        i += 1
+      }
+      nullify(from = i)
+      end = end_-(removals)
+    } else {
+      var i = idx + removals - 1
+      while(i - removals >= 0) {
+        set(i, get(i - removals))
+        set(i - removals, null)
+        i -= 1
+      }
+      nullify(until = i + 1)
+      start = start_+(removals)
+    }
+  }
+
+  override def remove(idx: Int) = {
+    val elem = this(idx)
+    remove(idx, 1)
+    elem
+  }
+
+  /**
+    *
+    * @param resizeInternalRepr If this is set, resize the internal representation to reclaim space once in a while
+    * @return
+    */
+  def removeFirst(resizeInternalRepr: Boolean = false): Option[A] = {
+    if (isEmpty) {
+      None
+    } else {
+      val elem = array(start)
+      array(start) = null
+      start = start_+(1)
+      if (resizeInternalRepr && 2*size < mask) resize(size)
+      Some(elem.asInstanceOf[A])
+    }
+  }
+
+  /**
+    *
+    * @param resizeInternalRepr If this is set, resize the internal representation to reclaim space once in a while
+    * @return
+    */
+  def removeLast(resizeInternalRepr: Boolean = false): Option[A] = {
+    if (isEmpty) {
+      None
+    } else {
+      end = end_-(1)
+      val elem = array(end)
+      array(end) = null
+      if (resizeInternalRepr && 2*size < mask) resize(size)
+      Some(elem.asInstanceOf[A])
+    }
+  }
+
+  override def reverse = foldLeft(new ArrayDeque[A](initialSize = size))(_.prependAssumingCapacity(_))
+
+  override def sizeHint(hint: Int) = if (hint >= mask) resize(hint + 1)
```

Comment:
No, I'm certainly not going to because that code is GPLed and this code is not.  If you've been using that as an extensive reference, I'm afraid we'll have to reimplement this all from scratch :(